src/renderer/html_handlebars/search.rs

```diff
@@ -0,0 +1,221 @@
+                    let doc_ref = make_doc_ref(&anchor_base, section_id);
+
+                    index.add_doc(&doc_ref, &[&heading, &body, &breadcrumbs.join(" » ")]);
+                    section_id = None;
+                    heading.clear();
+                    body.clear();
+                    breadcrumbs.pop();
+                }
+
+                in_header = true;
+            }
+            Event::End(Tag::Header(i)) if i <= max_section_depth => {
+                in_header = false;
+                section_id = Some(utils::id_from_content(&heading));
+                breadcrumbs.push(heading.clone());
+            }
+            Event::Start(Tag::BlockQuote)
+            | Event::End(Tag::BlockQuote)
+            | Event::Start(Tag::CodeBlock(_))
+            | Event::End(Tag::CodeBlock(_))
+            | Event::Start(Tag::Header(_))
+            | Event::End(Tag::Header(_))
+            | Event::End(Tag::Paragraph)
+            | Event::Start(Tag::Rule)
+            | Event::End(Tag::TableHead)
+            | Event::End(Tag::TableCell)
+            | Event::Start(Tag::List(_))
+            | Event::End(Tag::Item)
+            | Event::Start(Tag::Image(..))
+            | Event::End(Tag::Image(..))
+            | Event::SoftBreak
+            | Event::HardBreak => {
+                // Insert spaces where HTML output would usually seperate text
+                // to ensure words don't get merged together
+                if in_header {
+                    heading.push(' ');
+                } else {
+                    body.push(' ');
+                }
+            }
+            Event::Start(Tag::FootnoteDefinition(name)) => {
+                let number = footnote_numbers.len() + 1;
+                footnote_numbers.entry(name).or_insert(number);
+            }
+            Event::Start(_) | Event::End(_) => {}
+            Event::Text(text) => {
+                if in_header {
+                    heading.push_str(&text);
+                } else {
+                    body.push_str(&text);
+                }
+            }
+            Event::Html(html) | Event::InlineHtml(html) => {
+                let textified = utils::remove_html_tags(&html);
+                body.push_str(&textified);
+            }
+            Event::FootnoteReference(name) => {
+                let len = footnote_numbers.len() + 1;
+                let number = footnote_numbers.entry(name).or_insert(len);
+                body.push_str(&format!("[{}]", number));
+            }
+        }
+    }
+
+    if heading.len() > 0 {
+        // Make sure the last section is added to the index
+        let doc_ref = make_doc_ref(&anchor_base, section_id);
+
+        index.add_doc(&doc_ref, &[&heading, &body, &breadcrumbs.join(" » ")]);
+    }
+
+    Ok(())
+}
+
+/// Converts the index and search options to a JSON string
+fn write_to_json(index: Index, search_config: &Search) -> Result<String> {
```

Comment:
Do you have access to any "known good" versions of the generated JSON? It's probably a good idea to add a test or two to make sure we're able to generate valid JSON, that way we can also prevent accidental regressions in the future if someone needs to come back and change things.

Reply:
It's difficult to have any significant test for the entire search index. Tiny changes in the source material change the index drastically, and it's so large that it's difficult to determine if any change matters at all. If you have suggestions, I'm all ears.

Reply:
Hmm, you make a good point... What I usually do is generate a small example and store that as a file, giving it a quick look through to make sure it seems alright. Then every time after that you can have a test which uses the same input and check the rendered document is the same as our known good copy.

I have a similar issue at work where it's not really possible to manually construct an object and check each of its elements are the same every time. For that I've got a `create_fixture()` function which looks somewhat like this:

```rust
/// Regenerates the test fixture (pre-nested list of cutlines).
///
/// # Note
///
/// You'll need to run `cargo test` manually here, if tests are being run via
/// `cargo watch` and you try to regenerate the test fixture we'll get midway
/// through writing out the fixture before `cargo watch` restarts the process.
/// Leaving a mangled fixture file.
const GENERATE_FIXTURE: bool = false;


// loads of tests


fn get_fixture() -> Vec<Cutline> {
    if cfg!(windows) && GENERATE_FIXTURE {
        let inputs = create_known_dummy_inputs();
        let thing = generate_output(&inputs).unwrap();

        let fixture_path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/data/fixture.json");
        let mut f = ::std::fs::File::create(&fixture_path).unwrap();
        serde_json::to_writer_pretty(&mut f, &thing).unwrap();

        nested
    } else {
        let jason = include_str!("data/fixture.json");
        serde_json::from_str(jason).expect("Unable to deserialize the fixture")
    }
}
```

Reply:
Ah that's a good solution! One of the issues I was running into is that rendering dummy_book is different when you're in a test (due to env vars?) so it's annoying to make fixtures.